## PennCNV

### Installation

To install PennCNV, please follow the detailed instructions (including trouble shooting) at the [page](http://penncnv.openbioinformatics.org/en/latest/user-guide/install/). For more information about PennCNV, please refer to their original [PennCNV website](http://penncnv.openbioinformatics.org/en/latest/).

After installation, set up environment variable PENNCNV: `export PENNCNV='/path/to/penncnv'`

### Analysis workflow

Note: 

- PennCNV was originally designed to sequentially analyze one sample at a time. Please refer to [PennCNV website](http://penncnv.openbioinformatics.org/en/latest/) for how to perform a sequential analysis. Here, we provide scripts to run the analysis on multiple samples in parallel via job submitting system (one sample per job) in a cluster environment. 

- In the following steps (2) and (3), the scripts regarding job submission embraced by "##<<<... ##>>>..." in the scripts need to be specified by the users based on the system the users are using.

We run PennCNV analysis with the following 5 steps:

(1) Prepare SNP.pfb and SNP.gcmodel files

- compile pfb (population frequency of B allele) file
```sh
perl ${PENNCNV}/bin/compile_pfb.pl \
-listfile ${wk_dir}/list_pfb.txt \
-snpposfile ${wk_dir}/SNP_pos.txt \
-output ${wk_dir}/SNP.pfb
```

- compile gcmodel file for GC content ajdustment
## please refer to PennCNV documents about how to prepare gc5Base_hg19.txt.sorted
```sh
perl ${PENNCNV}/bin/cal_gc_snp.pl \
${wk_dir}/gc5Base_hg19.txt.sorted \
${wk_dir}/SNP.pfb \
-output ${wk_dir}/SNP.gcmodel
```

(2) Run PennCNV for each sample in parallel (through job submitting system on cluster)

Note: 

- In `step.2.run.PennCNV.jobs.R`, The scripts regarding job submission embraced by "##<<<... ##>>>..." need to be specified based on your system.

- For more information about CNV calling by PennCNV, please refer to the [page](http://penncnv.openbioinformatics.org/en/latest/user-guide/test/).

```sh 
Rscript ${WKDIR}/01_initial_call/run_PennCNV/step.2.run.PennCNV.jobs.R \
--penncnv ${PENNCNV} \                                  ## direct to ${PENNCNV}/bin/detect_cnv.pl
--data ${WKDIR}/01_initial_call/run_PennCNV/data \      ## generated with finalreport_to_PennCNV.pl
--wkdir ${WKDIR}/01_initial_call/run_PennCNV/results \  ## output directory
--pfb ${WKDIR}/01_initial_call/run_PennCNV/data_aux/SNP.pfb \
--gcmodel ${WKDIR}/01_initial_call/run_PennCNV/data_aux/SNP.gcmodel \
--hmm ${PENNCNV}/lib/hhall.hmm
```

When the analysis is completed, there will be subfolders named after sample IDs, each for one sample respectively, created in the directory `${WKDIR}/01_initial_call/run_PennCNV/results/res`. Within each sample subfolders, two files will be generated:
- `<Sample_ID>.log`: log file generated by `detect_cnv.pl`, information from which will be retrieved to generated sample-level summary statistics (see step (5) below). 
- `<Sample_ID>.rawcnv`: raw CNV calls made by `detect_cnv.pl`.

(3) Check job status and resubmit failed jobs

Note: In `step.3.check.PennCNV.jobs.R`, The scripts regarding job submission embraced by "##<<<... ##>>>..." need to be specified based on your system.

```sh
Rscrip ${WKDIR}/01_initial_call/run_PennCNV/step.3.check.PennCNV.jobs.R \
--penncnv ${PENNCNV} \                                  ## direct to ${PENNCNV}/bin/detect_cnv.pl
--data ${WKDIR}/01_initial_call/run_PennCNV/data/ \     ## generated with finalreport_to_PennCNV.pl
--wkdir ${WKDIR}/01_initial_call/run_PennCNV/results \  ## output directory
--pfb ${WKDIR}/01_initial_call/run_PennCNV/data_aux/SNP.pfb \
--gcmodel ${WKDIR}/01_initial_call/run_PennCNV/data_aux/SNP.gcmodel \
--hmm ${PENNCNV}/lib/hhall.hmm
```
This step will screen if the jobs submitted for each sample in step (2) are successfully completed and resubmit failed jobs if there is any.


(4) Combine PennCNV results (.rawcnv and .log files) from each sample
```sh
perl ${WKDIR}/01_initial_call/run_PennCNV/step.4.combine.PennCNV.res.pl \
--in_dir ${WKDIR}/01_initial_call/run_PennCNV/results/res \
--out_dir ${WKDIR}/01_initial_call/run_PennCNV/results
```
This script screens `.log` and `.rawcnv` files for all samples generated in steps (2) and (3), and combines them. When this step is completed, there will be two files generated in the directory `${WKDIR}/01_initial_call/run_PennCNV/results`:

- `CNV.PennCNV.log`: combined log file of the `.log` files from all samples.

- `CNV.PennCNV.rawcnv`: combined raw CNV calls of the `.rawcnv` files from all samples.

(5) Merge closely adjacent CNVs and generate final results
```sh
Rscript ${WKDIR}/01_initial_call/run_PennCNV/step.5.clean.PennCNV.res.R \
--penncnv ${PENNCNV} \                                  ## direct to installation directory ${PENNCNV}
--input ${WKDIR}/01_initial_call/run_PennCNV/results \
--pfb ${WKDIR}/01_initial_call/run_PennCNV/data_aux/SNP.pfb \
```

This script is a wrapper to run three perl scripts in PennCNV package will be used:

- `${PENNCNV}/bin/clean_cnv.pl`: merge adjacent CNVs which are close to each other. Please refer to the [page](http://penncnv.openbioinformatics.org/en/latest/user-guide/annotation/#merging-adjacent-cnv-calls) for details.

- `${PENNCNV}/bin/convert_cnv.pl`: convert the CNV calls in `.rawcnv` format to the tab-delimited table.

- `${PENNCNV}/bin/filter_cnv.pl`: extract sample-level summary statistics from log file.

When the analysis is completed, you will find two files, which will be used by ensembleCNV, in the directory `${WKDIR}/01_initial_call/run_PennCNV/results`:

- `CNV.PennCNV_new.txt`: CNV calls of all samples.
- `CNV.PennCNV_qc_new.txt`: sample-level summary statistics.

